Rdar 27397701 synthesized get set typechecking

graydon opened the pull request:
Addresses redundant typechecking of getter/setter function bodies outside of the primary file.
This is responsible for unnecessary quadratic work when doing a multi-file compilation.
PR also includes more scale-testing machinery, and a regression test that uses it.

Resolves [SR-2901](https://bugs.swift.org/browse/SR-2901).
Resolves rdar://27397701

graydon commented:
@swift-ci please test

huonw commented on include/swift/Basic/Statistic.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "llvm/ADT/Statistic.h"
+
+#define SWIFT_FUNC_STAT \
```
Does this need a `do { ... } while (0)` given how often braceless control flow is used, or are macro hacks like that not Swift/LLVM style?

jrose-apple replied:
No, that definitely would be preferred.

jrose-apple commented on lib/Sema/CodeSynthesis.cpp:
```diff
@@ -763,17 +767,25 @@ void swift::addTrivialAccessorsToStorage(AbstractStorageDecl *storage,
   synthesizeTrivialGetter(getter, storage, TC);
-  TC.typeCheckDecl(getter, true);
-  TC.typeCheckDecl(getter, false);
+  if (storage->getDeclContext()->isModuleScopeContext()) {
```
I don't see why this is the right check. What's different about members? Why can't I just validate accessors at module scope?

graydon replied:
Without that check, various resilience tests fail when emitting global getters (compiler crashes). I imagine this is due to unconditional emission of those getters, w/o being driven by any other references.

Seems to work when I tighten the condition to "only when in resilient mode", so I'll try that (and add a comment to that extent).

DougGregor commented:
Looks like we can't rely on numpy being around

graydon commented:
Split scale-test and stats helpers into separate PRs (#5607 and #5608 respectively) that this one depends on. Also fixed numpy and macro definition problems in them.

graydon commented:
Updated to unconditionally typecheck all accessors-with-bodies in decl visitor (that's applied to all SF-local decls) rather than trying to switch behaviour for resilient globals.

PR still depends on #5607 and #5608 to land first.

DougGregor commented:
This looks fantastic, thanks!

graydon commented:
@swift-ci please smoke test and merge

graydon commented:
@swift-ci please smoke test linux and merge

graydon commented:
@swift-ci please smoke test linux
